// @mintlify/common/dist/mdx/getMDXOptions.js
import { rehypeCodeBlocks, rehypeLayouts, rehypeListRoles, rehypeMdxExtractExamples, rehypeMdxInjectEndpoint, rehypeRawComponents, rehypeZoomImages, remarkFrames, remarkMdxInjectSnippets, remarkMdxRemoveJs, remarkRemoveImports, remarkTableOfContents, } from './plugins/index.js';
import rehypeMdxRemoveUnknownJsx from './plugins/rehype/rehypeMdxRemoveUnknownJsx/index.js';
import { remarkMermaid } from './plugins/remark/remarkMermaid.js';
export const getMDXOptions = ({ data, remarkPlugins = [], rehypePlugins = [], }) => {
    return {
        remarkPlugins: [
            [remarkMdxInjectSnippets, data.snippetTreeMap],
            remarkMdxRemoveJs,
            remarkFrames,
            remarkTableOfContents,
            remarkRemoveImports,
            remarkMermaid,
            ...remarkPlugins,
        ],
        rehypePlugins: [
            [
                rehypeMdxRemoveUnknownJsx,
                {
                    allowedComponents: data.allowedComponents,
                },
            ],
            rehypeCodeBlocks,
            rehypeListRoles,
            rehypeMdxExtractExamples,
            [rehypeMdxInjectEndpoint, data.pageMetadata, data.mintConfig || {}],
            rehypeRawComponents,
            rehypeLayouts,
            rehypeZoomImages,
            ...rehypePlugins,
        ],
        format: 'mdx',
        useDynamicImport: true,
    };
};
